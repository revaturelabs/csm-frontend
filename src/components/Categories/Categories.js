import React, { useEffect } from "react";
import { Menu, List, Card } from "semantic-ui-react";
import { useSelector, useDispatch } from "react-redux";
import Search from "../Search/Search.js";
import categoryService from "../../services/categories.service";

const Categories = (props) => {
  const dispatch = useDispatch();
  const categories = useSelector((state) => state.swotReducer.categories);
  const displayCategories = useSelector(
    (state) => state.swotReducer.displayCategories
  );

  const cat = new categoryService();

  const listDrag = (event) => {
    event.dataTransfer.setData("text", 'NONE~'+event.target.id);
  };

  useEffect(() => {
    async function getCat() {
      const resp = await cat.getCategories();
      const otherId = resp.length + 1;
      const lst = [...resp.data, {categoryId: otherId, skillCategory: 'Other', active:'true' }]
      lst.sort();
      dispatch({
        type: "updateDisplayCategories",
        getDisplayCategories: lst,
      });
    }
    getCat();
  }, [dispatch]);

  return (
    <>
      <Menu style={{ height: "75vh", overflowY: "scroll" }} vertical fluid>
        <Menu.Item>
          <Search />
        </Menu.Item>
        <Menu.Item>
          <List as="ul" style={{ marginLeft: 0 }}>
            {displayCategories.map((category) => (
              <Card key={category.categoryId} raised={true}>
                <List.Item
                  id={category.skillCategory}
                  name="category"
                  draggable
                  onDragStart={listDrag}
                >
                  {category.skillCategory}
                </List.Item>
              </Card>
            ))}
          </List>
        </Menu.Item>
      </Menu>
    </>
  );
};

export default Categories;
