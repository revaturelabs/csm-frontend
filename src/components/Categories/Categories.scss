@import "../../styles/main";

.item.swot-category-search {
  background-color: $revature-blue !important;
  border-radius: 0 !important;
}

.ui.menu#categories::-webkit-scrollbar {
  width: 10px;
}

.ui.menu#categories {
  height: 85vh;
  overflow-y: scroll;
  -ms-overflow-style: none;
  scrollbar-width: 3px; 
}
