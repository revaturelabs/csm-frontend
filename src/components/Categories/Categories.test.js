// Component.test.js
import React from "react";
import ReactDOM from "react-dom";
import Enzyme, { shallow, mount, render, fireEvent } from 'enzyme';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { useSelector, useDispatch } from 'react-redux';
import Categories from './Categories';

// It can receive two more parameters, the second one is to specify a factory instead of the jest's automocking feature
jest.mock('react-redux');

describe("Search unit test", () => {
    let wrapper;
    let useEffect;
    let store;

    const mockUseEffect = () => {
        useEffect.mockImplementationOnce(f => f());
    };

    beforeEach(() => {
        /* mocking store */
        store = configureStore([thunk])({
            swotReducer: {
                categories: ["AWS", "Python", "JavaScript"],
                displayCategories: ["AWS", "JavaScript"]
            },
        });

        /* mocking useEffect */
        useEffect = jest.spyOn(React, "useEffect");
        mockUseEffect(); // 2 times
        mockUseEffect(); //

        /* mocking useDispatch on our mock store */
        jest
            .spyOn(ReactReduxHooks, "useDispatch")
            .mockImplementation(() => store.dispatch);

        /* shallow rendering */
        wrapper = shallow(<RecipeList store={store} />);
    });

    describe("on mount", () => {
        it("Dispatch search action to store.", () => {
            const actions = store.getActions();
            expect(actions).toEqual([{type: "SEARCH", query: "all" },
            { type: "SEARCH_SUCCESS", categories: ["AWS", "Python", "JavaScript"]}]);
        }
    });

  	it('Rendering component without props.', () => {
		const { useSelector } = require("react-redux");
		useSelector.mockImplementation((callback) => {
			return callback({
				swotReducer: {
					categories: [],
					displayCategories: ["AWS", "Python", "JavaScript"],
				},
			});
		});
		const component = shallow(<Categories/>);
		expect(component).toMatchSnapshot();
  	});

  	it('Rendering component with children.', () => {
		const { useSelector } = require("react-redux");
		useSelector.mockImplementation((callback) => {
			return callback({
				swotReducer: {
					categories: [],
					displayCategories: ["AWS", "Python", "JavaScript"],
				},
			});
		});
		const component = render(<Categories/>);
		expect(component).toMatchSnapshot();
  	});

  	it('Mounting, testing list item mapping, and dismounting the component.', () => {
        const { useSelector, useDispatch } = require("react-redux");
        spyOn(React, 'useEffect').mockImplementation(f => f());
		useSelector.mockImplementation((callback) => {
            dispatch({
                type: "updateDisplayCategories",
                getDisplayCategories: categories,
            })
        }, [dispatch]);
		const component = mount(<Categories/>);
		expect(component).toMatchSnapshot();
		component.unmount();
  	});


});
