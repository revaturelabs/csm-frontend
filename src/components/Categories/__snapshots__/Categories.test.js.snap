// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search unit test on mount Rendering component with children. 1`] = `
<div
  class="ui fluid vertical menu"
  style="height:75vh;overflow-y:scroll"
>
  <div
    class="item"
  >
    <div
      class="ui fluid icon input"
    >
      <input
        id="search"
        placeholder="Search..."
        type="text"
      />
      <i
        aria-hidden="true"
        class="search icon"
      />
    </div>
  </div>
  <div
    class="item"
  >
    <ul
      class="ui list"
      role="list"
      style="margin-left:0"
    >
      <div
        class="ui raised card"
      >
        <div
          class="item"
          draggable="true"
          id="AWS"
          name="category"
          role="listitem"
        >
          AWS
        </div>
      </div>
      <div
        class="ui raised card"
      >
        <div
          class="item"
          draggable="true"
          id="Python"
          name="category"
          role="listitem"
        >
          Python
        </div>
      </div>
      <div
        class="ui raised card"
      >
        <div
          class="item"
          draggable="true"
          id="JavaScript"
          name="category"
          role="listitem"
        >
          JavaScript
        </div>
      </div>
    </ul>
  </div>
</div>
`;

exports[`Search unit test on mount Rendering component without props. 1`] = `
<Fragment>
  <Menu
    fluid={true}
    style={
      Object {
        "height": "75vh",
        "overflowY": "scroll",
      }
    }
    vertical={true}
  >
    <MenuItem>
      <Search />
    </MenuItem>
    <MenuItem>
      <List
        as="ul"
        style={
          Object {
            "marginLeft": 0,
          }
        }
      >
        <Card
          key="0"
          raised={true}
        >
          <ListItem
            draggable={true}
            id="AWS"
            name="category"
            onDragStart={[Function]}
          >
            AWS
          </ListItem>
        </Card>
        <Card
          key="1"
          raised={true}
        >
          <ListItem
            draggable={true}
            id="Python"
            name="category"
            onDragStart={[Function]}
          >
            Python
          </ListItem>
        </Card>
        <Card
          key="2"
          raised={true}
        >
          <ListItem
            draggable={true}
            id="JavaScript"
            name="category"
            onDragStart={[Function]}
          >
            JavaScript
          </ListItem>
        </Card>
      </List>
    </MenuItem>
  </Menu>
</Fragment>
`;
